Project resources extension

cmurphy opened the pull request:
https://github.com/rancher/rancher/issues/40140
 
Add support for a kubernetes API extension to enable listing resources
by one or more projects or namespaces.

From a Kubernetes perspective, the API is accessible at endpoints like

/apis/resources.project.cattle.io/v1alpha1/{resource}
/apis/resources.project.cattle.io/v1alpha1/namespaces/{projectID}/{resource}
/apis/resources.project.cattle.io/v1alpha1/namespaces/cattle-unscoped/{resource}

From a Steve perspective, the API is accessible at

/v1/resources.project.cattle.io.{group and resource}

For example,

/v1/resources.project.cattle.io.pods

or

/v1/resources.project.cattle.io.apps.deployments

To filter the results by a set of projects or namespaces, use the
`fieldSelector` query with the `prjorns` key, for example:

/v1/resources.project.cattle.io.pods?fieldSelector=projectsornamespaces={projec1},{project2},{namespace3}

To exclude projects or namespaces from the result, use the ! operator,
for example:

/v1/resources.project.cattle.io.pods?fieldSelector=projectsornamespaces!={project1},{project2},{namespace3}

This change adds the following:

1. Sets up a refresher to keep track of changing API resources (from
   added/removed CRDs and APIServices), so that namespaced API resources
   can be registered at this endpoint.

2. Applies the APIService pointing to the rancher or
   cattle-cluster-agent Service.

3. Sets up a handler for a discovery endpoint that responds with all the
   namespaced resources found in (1).

4. Sets up handlers for the global, project-scoped, and unscoped
   endpoints.

   In general, only admins have access to the global resource endpoint
   for any resource. Steve abstracts this by retrieving and collating
   resources from sets of namespaces that a standard user has access to.
   This means that Steve must be able to make requests to endpoints that
   have /namespaces/{namespace} in them, and a real namespace must exist
   representing the project, along with a role binding in that namespace
   so that the kube API server will allow the user to access that
   endpoint. We also need a special non-project namespace for Steve to
   go to to look for resources in namespaces that don't belong to any
   project.

5. Adds functionality to the rbac managementuser handlers to
   automatically create and remove the namespace and rolebinding for the
   project.

cmurphy commented:
Integration tests at https://github.com/rancher/rancher/pull/40536

Reviewer notes: open to suggestions for naming things, I'm not sure about the term "collections" or the keyword "prjorns" but haven't been able to think of anything better.

For reference, the alternative to doing this in rancher is adding something like this to dteve: https://github.com/cmurphy/steve/commit/8571f06361250fa5d8f2efbc25ccae0cd10ed0ec

rmweir commented on pkg/api/steve/collections/collections_test.go:
```diff
@@ -0,0 +1,1378 @@
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/gorilla/mux"
```
Move this to other block.

cmurphy replied:
done

rmweir reviewed (changes requested):
I tried running the example `v1/collections.cattle.io.pods?fieldSelector=prjorns=system` locally but I'm getting back an incomplete response.

cmurphy commented on pkg/api/steve/collections/apiresources.go:
```diff
@@ -0,0 +1,172 @@
+		time.Sleep(queueRefreshDelay * time.Millisecond)
+		if err := a.refreshAll(); err != nil {
+			logrus.Errorf("[%s] failed to sync schemas, will retry: %v", apiServiceName, err)
+			a.toSync.Store(1)
```
Pattern I copied. I will need to look closer at this, I am sure that I've seen it retry to refresh but I'm not seeing how.

cmurphy commented on pkg/api/steve/collections/apiresources.go:
```diff
@@ -0,0 +1,172 @@
+				StorageVersionHash: apidiscovery.StorageVersionHash(group, version, r.Kind),
+			}
+			result = append(result, resource)
+			a.resourceMap[name] = resource
```
There's nothing that cleans it up. I'll add that. Any leftover resources will increase the size of the map but won't otherwise cause any issue, the authoritative list that the API server uses is `a.apiResources`, this map is used to get individual resources in `Get()`.

cmurphy replied:
Adding a reset for this.

cmurphy commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+	return service, err
+}
+
+func getDynamicClient(config *wrangler.Context) (dynamic.Interface, error) {
+	restConfig, err := config.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return dynamic.NewForConfig(restConfig)
+}
```
Ill look into the factory approach

cmurphy replied:
Tried to switch this to use the client factory, but it looks like lasso needs you to provide a typed object to marshal the result into, and we can't know the object type ahead of time. With the dynamic client, it returns an unstructured list with "kind" set, so it's ready to go.

As for the informer idea, this needs to be able to support every ListOption that the k8s API provides, and informers aren't that flexible.

cmurphy reviewed (commented):
> I tried running the example v1/collections.cattle.io.pods?fieldSelector=prjorns=system locally but I'm getting back an incomplete response. It looks like the apiserver runs into a panic due to a nil schema. The k8s example does appear to work for me.

Yes, sorry I should have included this in the description. This relies on being able to point the API service at either the rancher or cattle-cluster-agent service, and in dev mode that doesn't exist (it works fine on helm/docker runs or if you run this in the agent). This is what I used to workaround that: https://gist.github.com/cmurphy/5ebc80503f6a8cb362c7a3111302bc04

I can add that YAML to some kind of dev/examples directory, I haven't been able to think of a better way to automatically support it without this intervention.

KevinJoiner commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+	}
+	kind, err := h.apis.GetKindForResource(resource)
+	if err != nil {
+		return nil, err
```
wrap error

KevinJoiner commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+}
+
+// hasAccess checks the auth headers set by the kube API server for the user's access to the requested resource in the requested project.
+func (h *handler) hasAccess(r *http.Request, namespace string, resource schema.GroupVersionResource) bool {
```
Would we also want to handle impersonation requests?

KevinJoiner commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+	return resourceList.GetResourceVersion(), nil
+}
+
+func responseData(resource schema.GroupVersionResource, kind, resourceVersion string, items []unstructured.Unstructured) map[string]interface{} {
+	return map[string]interface{}{
+		"apiVersion": resource.GroupVersion().String(),
+		"kind":       kind,
+		"metadata": map[string]interface{}{
+			"resourceVersion": resourceVersion,
+		},
+		"items": items,
+	}
+}
+
+func returnResp(w http.ResponseWriter, resp map[string]interface{}) {
+	resourceJSON, err := json.Marshal(resp)
```
I think it will be safer to use an `unstructuredList` than defining our own map
```suggestion
func responseData(resource schema.GroupVersionResource, kind, resourceVersion string, items []unstructured.Unstructured) unstructured.UnstructuredList{
	 list := unstructured.UnstructuredList {
		Items: items,
	 }
	 list.SetAPIVersion(resource.GroupVersion().String())
	 list.SetKind(kind)
	 list.SetResourceVersion(resourceVersion)
	
	return list
}

func returnResp(w http.ResponseWriter, resp unstructured.UnstructuredList) {
	resourceJSON, err := json.Marshal(resp)
```

KevinJoiner commented on pkg/controllers/managementuser/rbac/project_handler.go:
```diff
@@ -173,3 +195,32 @@ func getDefaultAndSystemProjectsToNamespaces() (map[string][]string, error) {
+
+// createOrUpdateProjectNS creates a namespace representing the project in the downstream cluster
+// for the purpose of assign roles on it. This allows the collections.cattle.io API to work.
+func (p *pLifecycle) createOrUpdateProjectNS(project *v3.Project) error {
```
I like this additions 🥇

KevinJoiner reviewed (changes requested):
I didn't review the test, but I will come back and look at them.

MbolotSuse commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+func (h *handler) scopedHandler(w http.ResponseWriter, r *http.Request) {
+	logrus.Tracef("[%s] handling request %s\n", apiServiceName, r.URL.Path)
+	vars := mux.Vars(r)
+	project := vars["project"]
```
I'm a bit confused on how this value interacts with the `pjorns` value. It seems like we are offering two methods to filter - one that allows filtering for an arbitrary number of namespaces/projects, and one that offers filtering for one project. 

Given that the second filter includes the first, I'm not sure that having both of them makes sense in this context.

cmurphy replied:
See my response to Kevin and the updated doc comment, this complexity is a requirement for compatibility with steve.

MbolotSuse replied:
Resolving this, was explained offline.

MbolotSuse commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+			return
+		}
+	}
+	if op != "!" {
```
(This comment applies for the general filtering logic in the if/else block below).

I think this filtering logic is a bit complex and has enough edge cases that it will be tough to maintain over time.

What I would recommend instead is logic that does the following:
- First, get the project and all associated namespaces (you can use an indexer to map namespace to owning project so that this lookup is ok from a speed perspective)
- Then If you've requested a filter (through `pjorns`), apply as below:
  - If we are using a positive filter (i.e. =), only include namespaces from the project that meet the filter (i.e. only include stuff from the first step if it matches at least one entry)
  - If we are using a negative filter (i.e. !=) remove those namespaces from the project that meet the filter (i.e. remove stuff from the first step if it matches at least one entry)
  - Don't do any filtering on project (this is controversial, but I think that if you are accessing this endpoint for a project we should return stuff for the project and only for that project - in this context, filtering by project has no meaning). What this means in practice is that if we encounter a value passed in by `pjorns` that doesn't match one of our active namespaces, we ignore that parameter. 
  
I'm hoping that this will make things a bit more simple, but it's possible I'm missing a desired behavior, so please LMK what you think.

cmurphy replied:
> I think this filtering logic is a bit complex and has enough edge cases

Yes, the edge cases are why it is complex

> First, get the project and all associated namespaces

This is done on line 378 https://github.com/rancher/rancher/pull/40155/files#diff-ff7986e4e885b7d6b19f5c52f21f48e75b503e7acc41534744ddbc6b5bf68cbcR378

> If we are using a positive filter (i.e. =), only include namespaces from the project that meet the filter

lines 349 to 364

> If we are using a negative filter (i.e. !=) remove those namespaces from the project that meet the filter (i.e. remove stuff from the first step if it matches at least one entry)

365 to 377. We can't simply remove the namespaces from the result, because we are also looking for projects that don't match the endpoint. If I say /namespaces/project1?fieldSelector=prjorns!=project1 then we need to return an empty list.

> Don't do any filtering on project (this is controversial, but I think that if you are accessing this endpoint for a project we should return stuff for the project and only for that project - in this context, filtering by project has no meaning)

See my comments on steve compatibility. This does not return resources for projects that don't match the endpoint.

MbolotSuse reviewed (commented):
Some minor comments on the collections file, need to make a more detailed review in the future.

cmurphy commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+	apiServiceName            = version + "." + Group
+	groupVersion              = Group + "/" + version
+	version                   = "v1alpha1"
+	workers                   = int64(3)
```
Maybe something to consider for the future

cmurphy commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+	version                   = "v1alpha1"
+	workers                   = int64(3)
+	queryKey                  = "fieldSelector"
+	projectsOrNamespacesKey   = "prjorns"
```
I think this was meant for saving characters in the URL but in hindsight it made the implementation very complex, will look into having separate parameters.

cmurphy replied:
I played around with this and added the changes as a separate commit so it would be easy to see the difference. It actually adds some complexity in some annoying ways, 1) since this isn't the primary query key (`fieldSelector` is) then it needs a distinct separator that isn't `&` or `,` to separate the two keys, and 2) now there are two possible operators, equality/nonequality for projects and the same for namespaces. I updated the tests to make them pass but I would need to add more test cases to cover that second caveat better. Let me know what you think of this direction.

cmurphy replied:
To support something like `fieldSelector=projects=p1,p2+namespaces=n1,n2` which I attempted in f820be06f46e24616ef117f87ed6271de4d4d7f9 we have to decide what to do when selectors overlap and the operators are in conflict with each other. For example, if namespace n1 is part of project p1 and I have `fieldSelector=projects=p1+namespaces!=n1`, what's the result? Do projects override namespaces or vice versa, does != override = or vice versa? This isn't something the UI currently supports or has asked for. I'm leaning toward keeping things simple-ish by using just a single parameter for both.

MbolotSuse replied:
Given the complexity I would tend to agree with you. I think we get less value from this because projects are also namespaces now in the downstream - so there's no chance for overlap. 

I'm ok with using one parameter, but I'd like to request that:
- We make it clear in the docs/project-readme that filtering by `projNS=p-1234xz` will filter on the project if it's a project, and that you can't use the parameter to filter on the resources in the downstream namespace by the same name (i.e. you can't use this endpoint to get the resources in the downstream project namespace)
- We come up with something more clear than `pjorns` as the name for the parameter. But we can do this whenever since it's a simple name change. 

LMK if you disagree.

cmurphy replied:
Sounds good to me.

For naming, I'll go with `projectsornamespaces` unless anyone proposes something better.

cmurphy commented on pkg/api/steve/collections/collections.go:
```diff
@@ -0,0 +1,699 @@
+)
+
+const (
+	// Group is the APIGroup for thie APIService.
```
done

cmurphy commented:
Added a commit to have the PRTB controller generate clusterroles and rolebindings to allow access to the new API based on PRTBs and a new controller to automatically add mirror roles and rolebindings for manually-created rolebindings in namespaces that don't belong to projects, this way steve can look up resources by the namespace itself instead of a special "cattle-unscoped" meta-namespace, and all the SAR checks are removed.

Still todo - make a call on "prjorns" versus "projects" and "namespaces" as selector keys, and add more unit tests - there are edge cases where the current projects/namespaces approach doesn't work, and add integration tests for both the new RBAC model and the new selector API.

cmurphy commented:
I still need to update the integration tests so this may change if the integration tests uncover new issues, and I also think I'm going to change the term "collections" to something else, but for the most part this is ready to re-review.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	if rb == nil {
+		return nil, nil
+	}
+	if strings.HasSuffix(rb.GetName(), suffix) { // this rolebinding was created by this controller, don't recurse
```
Nit:
I'd rather we do this check with a label/annotation than based on naming syntax, especially since we intend this to be active (AFAICT) in downstream clusters.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+					return collectionRules, nil
+				}
+				if apiGroup == "*" {
+					apis := apis.List()
```
doing the check for the apiGroup == "*" in the inmost loop could lead to you re-calculating this multiple times. I would recommend checking in an outer loop instead. something like:

```go
for _, rule := range rules{
    var apiGroups []string
    var isAllGroups bool
    // if we have a "*", fan out the groups and make sure the resources know allGroups are to be considered
    for _, group := range rule.apiGroups{
        if group == "*"{
            isAllGroups = true
            apiGroups = apis.List()
            break
        }
        apiGroups = append(apiGroups, group)
    }
    for _, resource := range resources {
        // use already calculated groups here
    }
}
```

cmurphy replied:
Not sure if I'm understanding your example correctly. I think the inefficiency is built in to the way policy rules are structured, which is that there isn't a 1:1 mapping from resource to apigroup and you have to cross-match them all. So if I have

```
rules:
- apiGroups:
  - "networking.k8s.io"
  - "storage.k8s.io"
  resources:
  - "ingresses"
  - "storageclasses"
```
I have to check discovery for "networking.k8s.io.ingresses", "networking.k8s.io.storageclasses", "storage.k8s.io.ingresses" and "storage.k8s.io.storageclasses". So assuming none of the apigroups is "*", in the resources loop I still have to loop over the apigroups list again to try to find a resource/apigroup combination that exists.

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+					collectionRules[0].Resources = append(collectionRules[0].Resources, resource)
+					continue
+				}
+				if resource == "*" {
```
Nit: Since this represents the special case (all resources in group), I'd check for this first and then do the check on the non-* use case.

cmurphy replied:
The non-special case is shorter and reads easier to me being up front :woman_shrugging:

MbolotSuse replied:
Ok - resolving since this is based on subjective tastes and you are the PR author.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+					if len(collectionRules) > 0 {
+						sort.Strings(collectionRules[0].Resources)
+					}
+					return collectionRules, nil
```
I might be missing something, but this return feels premature to me. We could have a case where there are `*` resources on two api groups. This seems like it will only add the resources for one of those and return early.

cmurphy replied:
I think you're right, will fix

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

MbolotSuse reviewed (commented):
Review of the RBAC controller functionality. Still need to go over other parts, but thought I would submit what I have here so far.

cmurphy commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
```
I wrote it that way initially. The reason I split it into two loops is that I want to avoid calling `apis.List()` unnecessarily, since during startup it takes some time to populate. If I run into an error with `apis.List()`, I have to return an error and requeue the whole rolebinding, whether or not the rolebinding and its role reference even ultimately need to be mirrored. So this way is actually more efficient on the whole because the rolebindings that we don't need are popped from the work queue and we don't touch them again.

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

cmurphy commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
```
will fix

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

cmurphy commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	if len(validRules) == 0 {
+		return collectionRules, nil
+	}
+	if len(apis.List()) == 0 {
```
I'm calling it here for two reasons - I'm calling GetCollectionRules from two places, so if this check is here I only have to have it in one place instead of two, and for the reason given above where it might not be necessary to check it at all and delaying doing so gets more objects through the work queue faster.

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

cmurphy commented:
Updated to fix two issues I found while updating the integration tests (https://github.com/rancher/rancher/pull/40536), have not addressed Michael's comments yet.

cmurphy commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	apis              APIResourceWatcher
+}
+
+func (h *rbacHandler) OnChange(_ string, rb *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
```
The alternative to this controller is to keep a special dummy namespace with a free-for-all rolebinding and do the SAR check in code. That's still on the table, and one could argue it's much less invasive than duplicating all these rolebindings.

MbolotSuse replied:
Thanks for explaining a bit more, that definitely helps me to understand. If we want to take this route, I think you need 4 total controllers:

2 Binding controllers (RoleBinding and ClusterRoleBinding)
2 Role controllers (Role and ClusterRole).

You need the Role/ClusterRole controllers to handle the case of if a currently bound role changes permissions. For exmample, suppose I have RoleBinding `x` binds user `y` to role `z`. If I change the rules of role `z`, this controller will need to be re-enqueued to update the collections roles with the new rules. 

If you don't want to do a controller for this, you can instead use wrangler's related resource functionality to re-enqueue the binding controller for RoleBindings/ClusterRoleBindings which are attached to this role if you'd prefer that approach instead. The multiple controller approach has it's benefits - namely, you can separate out "which role gives what rule" logic into the role/clusterrole controller and let the bindings just handle bindings, but that's ultimately up to you.

> This is meant to address non-project RBAC, but your comment makes me realize that users could manually create rolebindings in project namespaces too and this would need to work on that, which it currently doesn't.

My initial thought here is that there's a lot of logic here for a more niche use case (users not using our RBAC CRDS in downstream clusters). Though I suppose that we are doing this so that the UI works in all circumstances, and without this the UI could fail to display resources in certain namespaces where they used default k8s rbac, which is probably not great either.

> The alternative to this controller is to keep a special dummy namespace with a free-for-all rolebinding and do the SAR check in code. That's still on the table, and one could argue it's much less invasive than duplicating all these rolebindings.

I don't think I understand this option very well. Would this option be to have a `cattle-unscoped` namespace that contains the `collections.cattle.io` bindings for non-project namespaces? Or would it be one dummy namespace for each real non-project namespace?

cmurphy replied:
Forgot to fully address some of these issues before pushing the latest change, will fix.


> I don't think I understand this option very well. Would this option be to have a cattle-unscoped namespace that contains the collections.cattle.io bindings for non-project namespaces? Or would it be one dummy namespace for each real non-project namespace?

The first, one dummy namespace for all non-project namespaces. So a namespace like `cattle-unscoped` would have a rolebinding like

```
subjects:
- kind: Group
  name: system:authenticated
roleRef:
  kind: ClusterRole
  name: cattle-collections-all
```

and a ClusterRole that permits '*' on 'collections.cattle.io'. This lets steve know that there is a "partition" that the user has access to so it can hit /namespaces/cattle-unscoped/{resource}.

> My initial thought here is that there's a lot of logic here for a more niche use case

Agreed, these are for very niche corner cases, but the dashboard doesn't care what is niche or not, it needs to work in all cases.

I think the question is, is it so niche that taking the slight performance penalty of doing this with the SAR check worth the reduction in complexity? With you pointing out that we may need four controllers here I may be leaning toward yes.

MbolotSuse replied:
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		return nil, err
+	}
+	if _, ok := namespace.GetLabels()[nslabels.ProjectIDFieldLabel]; ok { // this only applies to namespaces not in a project
```
As long as this is consistent and we have a reason to use the label over the annotation (which it sounds like we do), then I am ok with this. Resolving.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+			return nil, err
+		}
+		rules = clusterRole.Rules
+	} else {
```
Thanks for pointing that out, looks like you are correct. Looks like it was written with those type of fields to accommodate new RBAC types in the future, but we can address that if/when it arrives.

MbolotSuse commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -109,6 +121,16 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
+		return errors.Wrap(err, "couldn't ensure project roles")
+	}
+
 	for _, n := range namespaces {
```
I think you need code here to handle collection RBAC for the individual namespaces in the project too.

For example, say user `test-user` has a PRTB giving them owner on project `p-123xyz`. Namespaces `test-ns-1` and `test-ns-2` are in `test-proj`. If `test-user` uses this endpoint to filter on `projectsornamespaces=p-123xyz`, I would expect this all to work, since the PRTB handler has handled creating the collection RBAC to give `test-user` access to that endpoint. However, if the user filtered on `projectsornamespaces=test-ns-1`, I don't think that this would work, since the PRTB handler only creates RBAC for the project and the collections RBAC controllers skips namespaces which are in a project. Correct me if I am missing something though, it's possible that I'm missing a part of the RBAC.

cmurphy replied:
This should still work, because it's not the query that matters for RBAC, it's the path. So I could have /namespaces/p-123xyz/pods?fieldSelector=projectsornamespaces=test-ns-1, and as long as I have permission to get pods for project p-123xyz it should work.

cmurphy commented:
Discussed offline - the added overhead and complexity of managing individual PRTBs and also users' manually created rolebindings is not worth it yet. We can keep this simple to start and find ways to optimize in the future if we find we need to. The RBAC commit has been removed.

Other changes:
- added TRACE logs to the API resource watcher, after discussion about the possibility that the pattern I took inspiration from may have led to performance issues in certain cases. We don't have conclusive evidence that is the case, but added logs can help us keep an eye on this down the road.
- added a script to create the rancher service for dev clusters where that isn't automatically created, because the APIService depends on it
- changed the group name - "collections" kept making me think of debt collections, so I changed it to `resources.project.cattle.io`.

Integration tests also updated separately https://github.com/rancher/rancher/pull/40536

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,706 @@
+	for _, n := range projectsOrNamespaces {
+		_, err := h.namespaceCache.Get(n)
+		if !apierrors.IsNotFound(err) {
+			filteredNamespaces[n] = struct{}{}
```
fixed

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,706 @@
+			if _, ok := filteredNamespaces[ns.Name]; !ok {
+				finalNamespaces[ns.Name] = ns
+			}
+		}
```
done

cmurphy commented on pkg/controllers/managementuser/rbac/project_handler.go:
```diff
@@ -173,3 +195,32 @@ func getDefaultAndSystemProjectsToNamespaces() (map[string][]string, error) {
+		return nil
+	}
+	labels := namespace.GetLabels()
+	if _, ok := labels[projectresources.ParentLabel]; ok {
```
No, there's nowhere where it's flipping from true to false, it's just existing.

cmurphy commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -119,6 +123,11 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
 
+	err = p.ensureProjectResourcesRoleBinding(binding)
+	if err != nil {
+		return errors.Wrapf(err, "couldn't ensure project resources binding %s in %s", projectresources.RoleName, binding.ProjectName)
```
I'm only using it to stay consistent with the rest of the file and the rest of the package. I don't want to change it everywhere.

cmurphy commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -128,13 +137,62 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
+		if err != nil {
+			return errors.Wrapf(err, "failed to create rolebinding %s for namespace %s", projectresources.RoleName, projectID)
+		}
+	}
```
fixed

cmurphy commented on pkg/controllers/managementuser/rbac/prtb_handler_test.go:
```diff
@@ -186,3 +189,475 @@ func Test_manager_checkForGlobalResourceRules(t *testing.T) {
+		binding *v3.ProjectRoleTemplateBinding
+		cache   map[string]map[string]*rbacv1.RoleBinding
+		want    *rbacv1.RoleBinding
+		wantErr error
```
I like checking the error string because I've seen error messages that use the wrong variables or forget the interpolation variables so the message ends up with `%!s(MISSING)` in it.

KevinJoiner replied:
Fair enough, that makes sense.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+	if rb == nil {
+		return nil, nil
+	}
+	if strings.HasSuffix(rb.GetName(), suffix) { // this rolebinding was created by this controller, don't recurse
```
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+		}
+		return nil, err
+	}
+	if len(collectionRules) == 0 {
```
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+			return nil, err
+		}
+	}
+	if collectionRoleBinding.RoleRef.Kind == "ClusterRole" {
```
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/collections/rbac.go:
```diff
@@ -0,0 +1,315 @@
+
+const (
+	suffix                     = "-collections"
+	clusterRoleOwnerAnnotation = "authz.cluster.cattle.io/clusterrole-owner"
```
Resolving, approach changed, file no longer exists.

MbolotSuse commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,703 @@
+	results := make(chan unstructured.Unstructured)
+	eg, ctx := errgroup.WithContext(r.Context())
+	resourceCollection := make([]unstructured.Unstructured, 0)
+	latestResourceVersion := 0
```
I'm not sure that I quite understand the purpose of this value. Near as I can tell, we are using this to return the last resource version from the various collections we retrieved. So if the list in ns1 gives 2 and the list in ns2 gives 4, we set this value to 2.

This value seems like it would not have much meaning to me. We are returning it to the user as though it is the resourceVersion of this collection, but in reality each list is independent of one another. In the above example, I could update the coollection of ns1 (producing a version of 3), but we would still return 4 since 4 is still higher than 3, making the user think the collection hasn't changed. If I'm understanding the use of this value correctly, we should probably use a sum here rather than a max - that way if any of the underlying values change than our return value changes too. Though that could get pretty large - let me know what thoughts you have.,

MbolotSuse commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -128,13 +137,63 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
 	return p.reconcileProjectAccessToGlobalResources(binding, roles)
 }
 
+func (p *prtbLifecycle) ensureProjectResourcesRoleBinding(binding *v3.ProjectRoleTemplateBinding) error {
```
You are probably going to need similar logic in the CRTB handler, or users who have CRTB access to a cluster may not have the rbac necessary to use this endpoint. Correct me if you disagree.

cmurphy replied:
You are right, adding

cmurphy replied:
It turns out this wreaks havoc on steve when a user has a combination of CRTBs and PRTBs or CRTBs and regular rolebindings. If a user has a cluster-member CRTB and a project-owner PRTB and tries to list secrets, steve sees the generic CRTB and forwards the request to the global endpoint for secrets, even though the user only has access to secrets in one project. Will have to rethink this, and probably go back to creating custom roles and individual clusterroles and roles for different CRTBs and PRTBs.

MbolotSuse reviewed (changes requested):
A couple of mostly minor comments. Only major thing I think is missing is CRTB logic. I haven't yet reviewed the tests, will do so at a later time.

cmurphy commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -128,13 +137,63 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        projectresources.RoleName,
+				Namespace:   projectID,
+				Annotations: map[string]string{projectresources.AuthzAnnotation: bindingID},
```
Realized this is a bug, different PRTBs are sharing this rolebinding so they can't all have the same label.

cmurphy commented:
ready for review again

rmweir reviewed (changes requested):
I think QA should make sure that all the downstream resources are properly cleaned up once a cluster is deleted/removed from being imported, or projects/ns's are deleted.

cmurphy commented:
> I think QA should make sure that all the downstream resources are properly cleaned up once a cluster is deleted/removed from being imported, or projects/ns's are deleted.

There is coverage for deleting projects in https://github.com/rancher/rancher/pull/40536/commits/126889086b4d1d5250de9958559ab928bad9d27b#diff-93bcc46fcef3a0ed3aeccebe0af4891d0e65c0c0fe388e305ae0a9ddab2a432c

It's hard to add integration tests for removing imported clusters because the only one cluster is imported during setup.

rmweir commented:
> > I think QA should make sure that all the downstream resources are properly cleaned up once a cluster is deleted/removed from being imported, or projects/ns's are deleted.
> 
> There is coverage for deleting projects in [1268890#diff-93bcc46fcef3a0ed3aeccebe0af4891d0e65c0c0fe388e305ae0a9ddab2a432c](https://github.com/rancher/rancher/commit/126889086b4d1d5250de9958559ab928bad9d27b#diff-93bcc46fcef3a0ed3aeccebe0af4891d0e65c0c0fe388e305ae0a9ddab2a432c)
> 
> It's hard to add integration tests for removing imported clusters because the only one cluster is imported during setup.

Do you think it's very to have that covered by QA then?

rmweir commented on pkg/controllers/managementuser/rbac/project_handler.go:
```diff
@@ -173,3 +195,32 @@ func getDefaultAndSystemProjectsToNamespaces() (map[string][]string, error) {
+	if _, ok := labels[projectresources.ParentLabel]; ok {
+		return nil
+	}
+	labels[projectresources.ParentLabel] = "true"
```
Can we document then that these are reserved?

cmurphy replied:
Yes I think we can do that

rmweir reviewed (changes requested):
Left a comment [here](https://github.com/rancher/rancher/pull/40155#discussion_r1159102219) and [here](https://github.com/rancher/rancher/pull/40155#issuecomment-1498255018). If the answer to those two is yes, then LGTM. Otherwise lmk what you think.

cmurphy commented on pkg/controllers/managementuser/rbac/prtb_handler.go:
```diff
@@ -106,6 +106,11 @@ func (p *prtbLifecycle) syncPRTB(binding *v3.ProjectRoleTemplateBinding) error {
 	}
 
 	if err := p.m.ensureRoles(roles); err != nil {
+		if errors.Is(err, errNotReady) {
```
Then we would have two different implementations in the same file, it's more important to be consistent imo

KevinJoiner replied:
I understand your point, and we may just disagree, which is okay. In this instance, since the std library has a one-for-one replacement, I rather change the entire file if we care about consistency over adding new code with the deprecated package.  I am not willing to die on this hill so we can resolve and move on.

KevinJoiner replied:
also, this package has larger problems than using the `pkg/errors`

cmurphy replied:
I added a commit to change it for the entire module, now no one else will have to have this debate :)

cmurphy commented:
> Do you think it's very to have that covered by QA then?

Yes I'm not sure what point I was trying to make other than that I'm trying to automate as much of the QA work as I can, I'll make sure to add resource cleanup check to the QA template.

cmurphy reviewed (commented):
Going to look closer at https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#kubernetes-apiserver-client-authentication and https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#original-request-username-and-group in response to Kevin and Michael's notes.

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,703 @@
+// hasAccess checks the auth headers set by the kube API server for the user's access to the requested resource in the requested project.
+func (h *handler) hasAccess(r *http.Request, namespace string, resource schema.GroupVersionResource) bool {
+	var groups []string
+	if g, ok := r.Header["X-Remote-Group"]; ok {
```
https://kubernetes.slack.com/archives/C0EG7JC6T/p1680891474113059

tldr it's always X-Remote-User

KevinJoiner commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,708 @@
+	var groups []string
+	if g, ok := r.Header["X-Remote-Group"]; ok {
+		groups = g
+	}
```
I think extras are used outside of impersonation requests as well. One test could involve creating a request with `X-Remote-ExtraFoo` header and validating that the SAR request contains the correct extra values.

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+// The http.Server must have a TLS config with ClientAuth: tls.RequestClientCert set in order for the http.Request to have the client cert populated.
+// Normally mTLS would be handled at the network session layer automatically during the client connection when tls.RequireAndVerifyClientCert
+// is set on the Server's tls.Config, but since this server performs multiple functions and not all of them require mTLS, we handle it in the
+// application layer in this middleware.
```
Here's an alternative to this client tls verification: https://github.com/cmurphy/rancher/commit/891d0caf5e6416dc1441c56db633fc920f72eee9

Benefits of the alternative:
- relies on the stdlib to do client cert verification, less likely to introduce security holes
Drawbacks:
- loads the CA only on startup, if the CA is changed in the configmap it won't be used until rancher reloads
- it spreads out the logic to two places, here and pkg/tls, when it really only applies to this one endpoint. Most of rancher's tls has zero to do with the k8s API server's CA, so it's out of place to add that verification there.

cmurphy reviewed (commented):
There's still some flakiness in https://github.com/rancher/rancher/pull/40536 having to do with the rbac changes, but this is still good to review, I can debug and fix the flakiness in 40536 or another PR.

MbolotSuse commented on scripts/dev-setup.sh:
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Nit: 

Putting this in the `scripts` folder makes it accessible using make, which runs everything through dapper. Since this can't be used through dapper (at least AFAIK) and needs to be called directly, it might be good to put it into another folder which isn't usable from make. Another PR had a similar issue and they created a `dev-scripts` folder (hasn't merged yet: https://github.com/rancher/rancher/pull/41216), which might serve your purposes.

cmurphy replied:
Good find, will do that, thanks

MbolotSuse reviewed (commented):
Mostly focused on the RBAC. I think I'm mostly ok with the changes in that area, but I'm a bit uneasy about the use of `group` + `.` + `resource` to qualify a resource.

cmurphy commented on pkg/controllers/managementuser/rbac/crtb_handler.go:
```diff
@@ -61,16 +63,25 @@ func (c *crtbLifecycle) syncCRTB(binding *v3.ClusterRoleTemplateBinding) error {
+		if errors.Is(err, errNotReady) {
+			logrus.Debugf("not finished setting up project roles, retrying")
+			c.m.crtbs.Controller().EnqueueAfter(binding.Namespace, binding.Name, enqueueAfterPeriod)
+			return nil
```
It looks like ErrSkip is equivalent to nil

https://github.com/rancher/wrangler/blob/80fdf092be3b5e11ccddce9e85aaef142e5d1830/pkg/generic/controller.go#L20
https://github.com/rancher/lasso/blob/f2977a658e8e999c060aefb525fac7e74bbc01c8/pkg/controller/sharedhandler.go#L76

I'll change it just to be consistent

cmurphy commented on pkg/controllers/managementuser/rbac/handler_base.go:
```diff
@@ -140,28 +151,35 @@ type manager struct {
 	for _, rt := range rts {
-		if rt.External {
-			continue
+		if !rt.External {
```
The order is critical, `ensureProjectResourcesRoles` must come after the clusterroles are created, see the godoc for that method. Can add an extra comment here as well.

cmurphy commented on pkg/controllers/managementuser/rbac/project_handler.go:
```diff
@@ -81,6 +98,18 @@ func (p *pLifecycle) Remove(project *v3.Project) (runtime.Object, error) {
+		return nil, err
+	}
+	if _, ok := projectNamespace.Annotations[projectNamespaceAnnotation]; !ok { // let management controller handle local namespace projects
+		err = p.m.workload.Core.Namespaces("").Delete(project.Name, &metav1.DeleteOptions{})
```
Agreed, this type of change was part of the refactor in https://github.com/rancher/rancher/pull/40155/commits/5eaa6ad861ef578d81e7f2da7595b2399be6ea7f but I missed this file and inadvertently added more of it, will fix

cmurphy commented on pkg/controllers/managementuser/rbac/projectresources_handler.go:
```diff
@@ -0,0 +1,215 @@
+				_, ok := m.apis.Get(resource, apiGroup)
+				if ok {
+					if apiGroup != "" {
+						resource = apiGroup + "." + resource
```
First, it's unlikely (though not impossible) that we'd run into a resource like that. There's no built-in resource type like that, and you can't create a CRD with a `.` in the name (I can't find that documented, but if you try to do it you'll run into a validation message from the server). It's possible this could happen if someone is doing what we're doing, creating an APIService with such a name and group. You'd have to be an admin of the cluster to do this, so they'd be shooting themselves in the foot.

Second, this concatenation is just about finding the resource in the schema, which is already registered in 
https://github.com/rancher/rancher/pull/40155/files#diff-3b10d3f41f25da075a1775137aa0f88a76d011e728a97f248a4136daa81d8584R156 and comes from steve's schema pattern https://github.com/rancher/steve/blob/e6a8019546c176b440922f746e0bc9670b4f9c40/pkg/schema/converter/k8stonorman.go#L31. If overlapping groups is a problem, then the first problem is steve's schema registration, then the new API's schema registration, and then this rbac.

MbolotSuse replied:
Thanks for explaining.

cmurphy commented:
Added a commit to address feedback from Richard. Will squash that in and address Michael's feedback next week.

MbolotSuse commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+func init() {
+	metav1.AddToGroupVersion(paramScheme, metav1.SchemeGroupVersion)
+
+	var err error
+	selectorString := "!" + nslabels.ProjectIDFieldLabel
+	orphanNamespaceSelector, err = labels.Parse(selectorString)
+	if err != nil {
+		panic(fmt.Sprintf("programmer error: invalid selector %s", selectorString))
+	}
```
Nit: This might be more readable as:

```golang
requirement, err := labels.NewRequirement(nslabels.ProjectIDFieldLabel, selection.DoesNotExist, nil)
if err != nil {
	panic(fmt.Sprintf("programmer error: invalid selector %s", selectorString))
}
orphanNamespaceSelector = labels.NewSelector().Add(*requirement)
```

MbolotSuse commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+		}
+		// only include resources in namespaces specified in the selector, or NOT specified in the selector if the operator is !
+		_, ok := filteredNamespaces[ns.Name]
+		if (op != notOp && ok) || (op == notOp && !ok) {
+			finalNamespaces[ns.Name] = ns
+		}
```
Optionally, you can use another label selector above to filter out these resources so that you don't have to do this here. Something like the below should do the trick:

```golang
// orphanRequirement, in this example, is a requirement like this:
// labels.NewRequirement(nslabels.ProjectIDFieldLabel, selection.DoesNotExist, nil)
namespaceRequirement, err := labels.NewRequirement(corev1.LabelMetadataName, selection.NotIn, nsNames)
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
	return
}
selector = labels.NewSelector().Add(*orphanRequirement, *namespaceRequirement)

```

This would allow you to just list and convert, without needing to filter later.

MbolotSuse reviewed (commented):
Couple of comments on the handlers and cert validation, will keep making my way through this.

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+// Normally mTLS would be handled at the network session layer automatically during the client connection when tls.RequireAndVerifyClientCert
+// is set on the Server's tls.Config, but since this server performs multiple functions and not all of them require mTLS, we handle it in the
+// application layer in this middleware.
+func (a *authHandler) authenticateAPIServer(next http.Handler) http.Handler {
```
I think you're technically right, I don't see anything in kubernetes that sets a default for these values. kubeadm, kind, k3s and rke set one, I think in practice most providers should set it. I could log a warning if it's not set and default to open?

I noticed in the docs "Note: You can set this option to blank as --requestheader-allowed-names="". This will indicate to an extension apiserver that any CN is acceptable." which this doesn't currently handle, will fix that.

MbolotSuse replied:
> I could log a warning if it's not set and default to open?

I don't think we can default to open. Near as I can tell, we get the information on which user is making the request from the request from k8s. If we can't tell that the requesting user is k8s, any user could hit the service directly and just assert that they are an admin user (by including that users name in the headers), which would allow them to do things like list core secrets and effect cluster takeover. LMK if you are seeing something that could prevent that. 

I think if we want to use this in this fashion, we need to be sure that every cluster sets these args and be willing to error out if we can't get the configmap as desired.

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM([]byte(clientCA))
+	opts := x509.VerifyOptions{
+		Roots:         caCertPool,
```
> // CurrentTime is used to check the validity of all certificates in the
   // chain. If zero, the current time is used.

I think it's maybe used for tests, we don't have to override the time if it's using the real time.

cmurphy commented on pkg/api/steve/projectresources/handlers.go:
```diff
@@ -0,0 +1,853 @@
+	a := authHandler{
+		configMapCache: config.Core.ConfigMap().Cache(),
+	}
+	subRouter.Use(a.authenticateAPIServer)
```
Commit "Use original verbs for resource rules" takes care of this by adding `.Methods` to the handlers

cmurphy commented:
Addressed most feedback, going to take a careful look at the selector suggestions

MbolotSuse commented on dev-scripts/setup-service.sh:
```diff
@@ -0,0 +1,46 @@
+      value: "8443"
+    - name: LOCAL_PORT
+      value: "443"
+    image: marcnuri/port-forward
```
This image is hosted outside of the rancher org, hasn't been updated in ~2 years, and isn't pinned to a specific version. Would there be an alternative image that we can use that has a lower risk of takeover/malicious usage?

cmurphy replied:
That's part of why I kept this in a separate script instead of trying to automate it in rancher. I'll look for an alternative, or if you know of one I'll happily swap it out.

cmurphy commented on pkg/api/steve/projectresources/register.go:
```diff
@@ -0,0 +1,295 @@
+			Name:     RoleName,
+		},
+	}
+	err = a.apply.ApplyObjects(&apiService, &namespace, &role, &roleBinding)
```
I wrote this before that bug came out, it's just for convenience/following common rancher patterns. Happy to change it.

cmurphy replied:
FYI this is ~120 extra LOC to do this without wrangler apply.

cmurphy commented:
Latest changes replace the use of wrangler Apply and implement support for "Accept: application/json;as=Table" which steve needs in order to provide field headers for the UI. Still need to look at https://github.com/rancher/rancher/pull/40155#discussion_r1175796743 and related offline comments.
